syntax = "proto3";
import "common.proto";
package huawei.proto.control;

/*northern oriented rpc service interface*/
service SnapshotControl {
    rpc CreateSnapshot(CreateSnapshotReq) returns(CreateSnapshotAck){}
    rpc ListSnapshot(ListSnapshotReq) returns(ListSnapshotAck){}
    rpc QuerySnapshot(QuerySnapshotReq) returns(QuerySnapshotAck){}
    rpc RollbackSnapshot(RollbackSnapshotReq) returns(RollbackSnapshotAck){}
    rpc DeleteSnapshot(DeleteSnapshotReq) returns(DeleteSnapshotAck){}
    rpc DiffSnapshot(DiffSnapshotReq) returns(DiffSnapshotAck){}
    rpc ReadSnapshot(ReadSnapshotReq) returns(ReadSnapshotAck){}
}

message CreateSnapshotReq {
    SnapReqHead header = 1;
    string vol_name = 2;
    string snap_name = 3;
}

message CreateSnapshotAck {
    SnapAckHead header = 1;
}

message ListSnapshotReq {
    SnapReqHead header = 1;
    string vol_name = 2;
}

message ListSnapshotAck {
    SnapAckHead header = 1;
    repeated string snap_name = 2;
}

message QuerySnapshotReq {
    SnapReqHead header = 1;
    string vol_name = 2;
    string snap_name = 3;
}

message QuerySnapshotAck {
    SnapAckHead header = 1;
    SnapStatus snap_status = 2;
}

message RollbackSnapshotReq {
    SnapReqHead header = 1;
    string vol_name = 2;
    string snap_name = 3;
}

message RollbackSnapshotAck {
    SnapAckHead header = 1;
}

message DeleteSnapshotReq {
    SnapReqHead header = 1;
    string vol_name = 2;
    string snap_name = 3;
}

message DeleteSnapshotAck {
    SnapAckHead header = 1;
}

message DiffSnapshotReq {
    SnapReqHead header = 1;
    string vol_name = 2;
    string first_snap_name = 3;
    string last_snap_name = 4;
}

message ExtDiffBlock {
    uint64 off = 1;
    uint64 len = 2;
}

message ExtDiffBlocks {
    string vol_name  = 1;
    string snap_name = 2;
    repeated ExtDiffBlock diff_block = 3;
}

message DiffSnapshotAck {
    SnapAckHead header = 1;
    repeated ExtDiffBlocks diff_blocks = 2;
} 

message ReadSnapshotReq {
    SnapReqHead header = 1;
    string vol_name  = 2;
    string snap_name = 3;
    uint64 off       = 4;
    uint64 len       = 5;
}

message ReadSnapshotAck {
    SnapAckHead header = 1;
    repeated bytes data = 2;
}
